src/webgpu/shader/validation/functions/alias_analysis.spec.ts

```diff
@@ -0,0 +1,271 @@
+  { use: 'compound_assign_rhs', is_write: false },
+  { use: 'increment', is_write: true },
+  { use: 'binary_lhs', is_write: false },
+  { use: 'binary_rhs', is_write: false },
+  { use: 'unary_minus', is_write: false },
+  { use: 'bitcast', is_write: false },
+  { use: 'convert', is_write: false },
+  { use: 'function_arg', is_write: false },
+  { use: 'index_access', is_write: false },
+  { use: 'let_init', is_write: false },
+  { use: 'var_init', is_write: false },
+  { use: 'return', is_write: false },
+  { use: 'switch_cond', is_write: false },
+];
+
+function generateUse(condition: string, ref: string): string {
+  switch (condition) {
+    case 'no_access': {
+      return `{ let p = &*&${ref}; }`;
+    }
+    case 'assign': {
+      return `${ref} = 42;`;
+    }
+    case 'increment': {
+      return `${ref}++;`;
+    }
+    case 'compound_assign_lhs': {
+      return `${ref} += 1;`;
+    }
+    case 'compound_assign_rhs': {
+      return `{ var tmp : i32; tmp += ${ref}; }`;
+    }
+    case 'binary_lhs': {
+      return `_ = ${ref} + 1;`;
+    }
+    case 'binary_rhs': {
+      return `_ = 1 + ${ref};`;
+    }
+    case 'unary_minus': {
+      return `_ = -${ref};`;
+    }
+    case 'bitcast': {
+      return `_ = bitcast<f32>(${ref});`;
+    }
+    case 'convert': {
+      return `_ = f32(${ref});`;
+    }
+    case 'function_arg': {
+      return `_ = abs(${ref});`;
+    }
+    case 'index_access': {
+      return `{ var arr : array<i32, 4>; _ = arr[${ref}]; }`;
+    }
+    case 'let_init': {
+      return `{ let tmp = ${ref}; }`;
+    }
+    case 'var_init': {
+      return `{ var tmp = ${ref}; }`;
+    }
+    case 'return': {
+      return `{ return ${ref}; }`;
+    }
+    case 'switch_cond': {
+      return `switch(${ref}) { default { break; } }`;
+    }
+    default: {
+      unreachable(`Unhandled usage type`);
+    }
+  }
+}
+
+function shouldPass(aliased: boolean, a_use: Use, b_use: Use): boolean {
+  // Expect fail if the pointers are aliased and at least one of the accesses is a write.
+  // If either of the accesses is a "no access" then expect pass.
+  let should_pass = aliased ? !(a_use.is_write || b_use.is_write) : true;
+  should_pass = a_use.use === 'no_access' ? true : should_pass;
+  should_pass = b_use.use === 'no_access' ? true : should_pass;
+  return should_pass;
+}
+
+g.test('two_pointers')
+  .desc(`Test aliasing of two pointers passed to a function.`)
+  .params(u =>
+    u
+      .combine('address_space', ['private', 'function'] as const)
+      .combine('a_use', kUses)
+      .combine('b_use', kUses)
+      .combine('aliased', [true, false])
+      .beginSubcases()
+  )
+  .fn(t => {
+    const code = `
+${t.params.address_space === 'private' ? `var<private> va : i32; var<private> vb : i32;` : ``}
+
+fn callee(pa : ptr<${t.params.address_space}, i32>,
+          pb : ptr<${t.params.address_space}, i32>) -> i32 {
+  ${generateUse(t.params.a_use.use, `*pa`)}
+  ${generateUse(t.params.b_use.use, `*pb`)}
```

Comment:
... and then this becomes:

```suggestion
  ${kUses[t.params.a_use].gen(`*pa`)}
  ${kUses[t.params.b_use].gen(`*pb`)}
```